src/modules/ekf2/EKF/aid_sources/external_vision/ev_vel_control.cpp

```diff
@@ -232,10 +243,86 @@ void Ekf::controlEvVelFusion(const extVisionSample &ev_sample, const bool common
 	}
 }
 
+bool Ekf::fuseEvVelocity(estimator_aid_source3d_s &aid_src, const extVisionSample &ev_sample)
+{
+	if (ev_sample.vel_frame == VelocityFrame::BODY_FRAME_FRD) {
+
+		VectorState H;
+		estimator_aid_source1d_s current_aid_src;
+
+		for (uint8_t index = 0; index <= 2; index++) {
+			current_aid_src.timestamp_sample = aid_src.timestamp_sample;
+
+			if (index == 0) {
+				sym::ComputeEvBodyVelHx(_state.vector(), &H);
+
+			} else if (index == 1) {
+				sym::ComputeEvBodyVelHy(_state.vector(), &H);
+
+			} else {
+				sym::ComputeEvBodyVelHz(_state.vector(), &H);
+			}
+
+			const float innov_var = (H.T() * P * H)(0, 0) + aid_src.observation_variance[index];
+			const float innov = (_R_to_earth.transpose() * _state.vel - Vector3f(aid_src.observation))(index, 0);
+
+			updateAidSourceStatus(current_aid_src,
+					      ev_sample.time_us,				// sample timestamp
+					      aid_src.observation[index],			// observation
+					      aid_src.observation_variance[index],		// observation variance
+					      innov,						// innovation
+					      innov_var,    					// innovation variance
+					      math::max(_params.ev_vel_innov_gate, 1.f));	// innovation gate
+
+			if (!current_aid_src.innovation_rejected) {
+				fuseBodyVelocity(current_aid_src, current_aid_src.innovation_variance, H);
+
+			}
+
+			aid_src.innovation[index] = current_aid_src.innovation;
+			aid_src.innovation_variance[index] = current_aid_src.innovation_variance;
+			aid_src.test_ratio[index] = current_aid_src.test_ratio;
+			aid_src.fused = current_aid_src.fused;
+			aid_src.innovation_rejected |= current_aid_src.innovation_rejected;
+
+			if (aid_src.fused) {
+				aid_src.time_last_fuse = _time_delayed_us;
+			}
+
+		}
+
+		aid_src.timestamp_sample = current_aid_src.timestamp_sample;
+		return !aid_src.innovation_rejected;
+
+	} else {
+		return fuseVelocity(aid_src);
+	}
+}
+
 void Ekf::stopEvVelFusion()
 {
 	if (_control_status.flags.ev_vel) {
 
 		_control_status.flags.ev_vel = false;
 	}
 }
+
+void Ekf::resetVelocityToEV(const Vector3f &measurement, const Vector3f &measurement_var,
+			    const VelocityFrame &vel_frame)
+{
+	if (vel_frame == VelocityFrame::BODY_FRAME_FRD) {
```

Comment:
I think this needs a refactoring; It's not great that a function does two different things based on a parameter